File: scripts/gather-new-data/02_export_hydat_csv.R
# Author: Saeesh Mangwani
# Date: 2022-06-15

# Description: Exporting Hydat data to CSV format based on a provided
# date range

# ==== Libraries ====
library(DBI)
library(RPostgres)
library(rjson)
library(readr)
library(dplyr)
library(optparse)
library(lubridate)
source('scripts/help_funcs.R')

# ==== Initializing option parsing ====
option_list <-  list(
  make_option(c("-s", "--startdate"), type="character", default=(Sys.Date() - 30), 
              help="A year month combination indicating the start date for data to post to depth2water. Defaults to 30 days before today. [Default= %default]", 
              metavar="character"),
  make_option(c("-e", "--enddate"), type="character", default=(Sys.Date()), 
              help="A year month combination indicating the end date for data to post to depth2water. Defaults to the current date [Default= %default]", 
              metavar="character")
)

# Parse any provided options and store them in a list
opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)

# ==== Paths and global variables ====

# Database credentials
creds <- fromJSON(file = 'options/dbase_credentials.json')

# Filepaths for gathering functions
fpaths <- fromJSON(file='options/filepaths.json')

# Ensuring directory exists for holding posting data data
dir_check_create(fpaths$`daily-post-dir`)
# Creating a sub-directory for EC climate, and deleting any prior contents if
# they exist
out_dir <- file.path(fpaths$`daily-post-dir`, 'hydat')
if(dir.exists(out_dir)) unlink(out_dir, recursive=T)
dir_check_create(out_dir)

# ==== Opening database connection ====

# Opening database connection
conn <- dbConnect(RPostgres::Postgres(),
                  host = creds$host, dbname = creds$dbname,
                  user = creds$user, password = creds$password)

# ==== Reading data ====

# Flow data
query <- format_simple_query('bchydat', 'flow', 
                             date_col = 'Date', 
                             start_date = ymd(opt$startdate),
                             end_date = ymd(opt$enddate))
flow <- dbGetQuery(conn, query)

# Level data
query <- format_simple_query('bchydat', 'level', 
                             date_col = 'Date', 
                             start_date = ymd(opt$startdate),
                             end_date = ymd(opt$enddate))
level <- dbGetQuery(conn, query)

# Joining to a single table
hydat <- flow %>% 
  select(-Parameter, -Symbol) %>% 
  rename('flow' = Value) %>% 
  full_join(level %>% 
              select(STATION_NUMBER, Date, 'level' = Value, pub_status),
            by = c('STATION_NUMBER', 'Date')) %>%
  mutate(pub_status = case_when(
    is.na(pub_status.x) & is.na(pub_status.y) ~ NA_character_,
    is.na(pub_status.y) ~ pub_status.x,
    is.na(pub_status.x) ~ pub_status.y,
    T ~ pub_status.x
  )) %>% 
  select(-pub_status.x, -pub_status.y) %>% 
  # Converting pub-status to a boolean
  mutate(pub_status = ifelse(pub_status == 'Published', T, F))

# ==== Writing to CSV ====

# Writing if there are data, otherwise printing a nodata message
if(nrow(hydat) > 0){
  # Writing to csv
  write_csv(hydat, file.path(out_dir, 'hydat-daily.csv'))
}else{
  print(
    paste0(
      'No new data available for Hydat between ',
      as.character(opt$startdate),
      ' and ',
      as.character(opt$enddate),
      '. No CSV exported.'
    )
  )
}

# Closing database connection
dbDisconnect(conn)

File: scripts/gather-new-data/02_export_hydat_csv.py
# Author: Saeesh Mangwani
# Date: 02/08/2023

# Description: Exporting EC Climate data to CSV format based on a provided date range

# %% ===== Loading libraries =====
import os
import sys
from pathlib import Path
os.chdir(Path(__file__).parent.parent.parent)
sys.path.append(os.getcwd())
from json import load
from optparse import OptionParser
from datetime import datetime, timedelta
from numpy import where, NaN
import psycopg2
import pandas as pd


#%% Initializing option parsing
parser = OptionParser()
parser.add_option(
    "-s", "--startdate", 
    dest="startdate",
    default=(datetime.today() - timedelta(days=31)).strftime("%Y-%m-%dT00:00:00-00:00"),
    help="The start date of the date range for which data are being posted. Defaults to 31 days before today")
parser.add_option(
    "-e", "--enddate", 
    dest="enddate",
    default=datetime.today().strftime("%Y-%m-%dT00:00:00-00:00"),
    help="The end date of the date range for which data are being posted. Defaults to today")
(options, args) = parser.parse_args()

# %% ===== Paths and global variables =====

# Client credentials from JSON
creds = load(open('options/dbase_credentials.json',))

# Filepaths
fpaths = load(open('options/filepaths.json', ))

# Ensuring directory exists for holding posting data data
out_dir = fpaths['update-data-dir']
if not os.path.exists(out_dir):
    os.makedirs(out_dir)

# %% ===== Initializing database connection =====

# Database connection
conn = psycopg2.connect(
    host=creds['host'], 
    port=creds['port'],
    database=creds['dbname'],
    user= creds['user'],
    password=creds['password']
)
cursor = conn.cursor()

# %% ===== Initializing update options =====
start_date = options.startdate
end_date = options.enddate
# start_date = (datetime.today() - timedelta(days=331)).strftime("%Y-%m-%dT00:00:00-00:00")
# end_date =datetime.today().strftime("%Y-%m-%dT00:00:00-00:00")

# %% ==== Gathering update data ====

# Shared query options options
schema = 'bchydat'
datecol = 'Date'

# Flow data

# Building query
query = 'select * from {0}.{1} where "{2}" >= \'{3}\' and "{2}" <= \'{4}\''.format(
    schema,
    'flow',
    datecol,
    start_date,
    end_date
)

# Getting data
cursor.execute(query)
col_names = [i[0] for i in cursor.description]
flow = pd.DataFrame(cursor.fetchall(), columns=col_names)

# Level data
# Building query
query = 'select * from {0}.{1} where "{2}" >= \'{3}\' and "{2}" <= \'{4}\''.format(
    schema,
    'level',
    datecol,
    start_date,
    end_date
)

# Getting data
cursor.execute(query)
col_names = [i[0] for i in cursor.description]
level = pd.DataFrame(cursor.fetchall(), columns=col_names)

# %% Joining flow and level to a single table

# Selecting only relevant columns
flow = flow[['STATION_NUMBER', 'Date', 'Value','pub_status']]
flow.rename(columns={'Value': 'flow'}, inplace=True)

level = level[['STATION_NUMBER', 'Date', 'Value','pub_status']]
level.rename(columns={'Value': 'level'}, inplace=True)

# Full Joining
daily = flow.merge(level, how='outer', on=['STATION_NUMBER', 'Date'])

# Creating a synthesized pub-status column
daily['pub_status'] = where(
    daily['pub_status_x'].isna() & daily['pub_status_y'].isna(), NaN,
    where(daily['pub_status_x'].isna(), daily['pub_status_y'], daily['pub_status_x']))

# Dropping intermediate columns
daily.drop(columns=['pub_status_x', 'pub_status_y'], inplace=True)

# Converting pub status to boolean
daily['pub_status'] = [True if status == 'Published' else False for status in daily['pub_status']]

# %%  ==== Exporting to CSV ====
if daily.shape[0] == 0:
    print("No new data available for EC-Climate between {} and {}. No CSV exported".format(start_date, end_date))
else:
    print("Exporting data to CSV")
    daily.to_csv(out_dir + '/ecclimate-daily.csv', index=False)

# %%
